c++/src/kj/compat/http.c++

```diff
@@ -2594,12 +2653,205 @@ private:
+            // Therefore, a decompressor using `windowBits` of 8 could safely inflate a message
+            // that a zlib client compressed using `windowBits` = 9.
+            // https://bugs.chromium.org/p/chromium/issues/detail?id=691074
+            windowBits = -9;
+          }
+          int result = deflateInit2(
+              &ctx,
+              Z_DEFAULT_COMPRESSION,
+              Z_DEFLATED,
+              windowBits,
+              8,  // memLevel = 8 is the default
+              Z_DEFAULT_STRATEGY);
+          KJ_REQUIRE(result == Z_OK, "Failed to initialize compression context (deflate).");
+          break;
+        }
+        case Mode::DECOMPRESS: {
+          int windowBits = -config.inboundMaxWindowBits.orDefault(15);
+          // We use negative values because we want to use raw inflate.
+          int result = inflateInit2(&ctx, windowBits);
+          KJ_REQUIRE(result == Z_OK, "Failed to initialize decompression context (inflate).");
+          break;
+        }
+      }
+    }
+
+    ~ZlibContext() noexcept(false) {
+      switch (mode) {
+        case Mode::COMPRESS:
+          deflateEnd(&ctx);
+          break;
+        case Mode::DECOMPRESS:
+          inflateEnd(&ctx);
+          break;
+      }
+    }
+
+    KJ_DISALLOW_COPY(ZlibContext);
+
+    kj::Array<kj::byte> processMessage(kj::ArrayPtr<const byte> message,
+        kj::Maybe<size_t> maxSize = nullptr,
+        bool addNullTerminator = false) {
+      // If `this` is the compressor, calling `processMessage()` will compress the `message`.
+      // Likewise, if `this` is the decompressor, `processMessage()` will decompress the `message`.
+      //
+      // `maxSize` is only passed in when decompressing, since we want to ensure the decompressed
+      // message is smaller than the `maxSize` passed to `receive()`.
+      //
+      // If (de)compression is successful, the result is returned as a Vector, otherwise,
+      // an Exception is thrown.
+
+      ctx.next_in = const_cast<byte*>(reinterpret_cast<const byte*>(message.begin()));
+      ctx.avail_in = message.size();
+
+      kj::Vector<Result> parts(processLoop(maxSize));
+
+      size_t amountToAllocate = 0;
+      for (const auto& part : parts) {
+        amountToAllocate += part.size;
+      }
+
+      if (addNullTerminator) {
+        // Add space for the null-terminator.
+        amountToAllocate += 1;
+      }
+
+      kj::Array<kj::byte> processedMessage = kj::heapArray<kj::byte>(amountToAllocate);
+      size_t currentIndex = 0; // Current index into processedMessage.
+      for (const auto& part : parts) {
+        memcpy(&processedMessage[currentIndex], part.buffer.begin(), part.size);
+        // We need to use `part.size` to determine the number of useful bytes, since data after
+        // `part.size` is unused (and probably junk).
+        currentIndex += part.size;
+      }
+
+      if (addNullTerminator) {
+        processedMessage.back() = '\0';
+      }
+
```

Comment:
At the end of code like this I like to include an assertion that we didn't overrun or underrun the buffer:

`KJ_ASSERT(currentIndex == processedMessage.size());`

Reply:
Will keep that in mind, sounds like an easy way to prevent headaches: https://github.com/capnproto/capnproto/pull/1505/commits/e6972f939854ae79daa8869ff63a1e91f48b469d